src/docfx/lib/git/GitCommitProvider.cs

```diff
@@ -1,412 +1,74 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
-
 using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
-using System.Diagnostics;
 using System.IO;
-using System.Linq;
-using System.Runtime.InteropServices;
-using System.Threading;
 using System.Threading.Tasks;
 
-using static Microsoft.Docs.Build.LibGit2;
-
-#pragma warning disable CA2002 // Do not lock on objects with weak identity
-
 namespace Microsoft.Docs.Build
 {
     internal sealed class GitCommitProvider : IDisposable
     {
-        private const int MaxCommitCacheCountPerFile = 10;
-
-        private readonly string _repoPath;
-        private readonly string _cacheFilePath;
-
-        // Commit history and a lookup table from commit hash to commit.
-        // Use `long` to represent SHA2 git hashes for more efficient lookup and smaller size.
-        private readonly ConcurrentDictionary<string, Lazy<(List<Commit>, Dictionary<long, Commit>)>> _commits;
-
-        // Intern path strings by given each path segment a string ID. For faster string lookup.
-        private readonly ConcurrentDictionary<string, int> _stringPool = new ConcurrentDictionary<string, int>();
-
-        // A giant memory cache of git tree. Key is the `long` form of SHA2 tree hash, value is a string id to git SHA2 hash.
-        private readonly ConcurrentDictionary<long, Dictionary<int, git_oid>> _trees
-                   = new ConcurrentDictionary<long, Dictionary<int, git_oid>>();
+        private readonly ConcurrentDictionary<string, Repository> _repositoryByFolder = new ConcurrentDictionary<string, Repository>();
+        private readonly ConcurrentDictionary<string, Lazy<Task<(FileCommitProvider provider, Repository repo)>>> _commitProviders = new ConcurrentDictionary<string, Lazy<Task<(FileCommitProvider provider, Repository repo)>>>();
```

Comment:
nit: feels like `_commitProviders` is a superset of `_repositoryByFolder`, maybe one is enough?

Reply:
i would like to split them, make them more clear to me.

---
In reply to: [237933739](https://github.com/dotnet/docfx/pull/3705#discussion_r237933739) [](ancestors = 237933739)